contracts/puppeteer-native/src/contract.rs

```diff
@@ -0,0 +1,531 @@
+                let owner = cw_ownable::get_ownership(deps.storage)?;
+                to_json_binary(&owner).map_err(ContractError::Std)
+            }
+        },
+        QueryMsg::Config {} => query_config(deps),
+        QueryMsg::Transactions {} => query_transactions(deps),
+    }
+}
+
+fn query_config(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(to_json_binary(&config)?)
+}
+
+fn query_transactions(deps: Deps<NeutronQuery>) -> ContractResult<Binary> {
+    let transfers: Vec<Transfer> = RECIPIENT_TRANSFERS.load(deps.storage)?;
+    Ok(to_json_binary(&transfers)?)
+}
+
+fn query_delegations(deps: Deps<NeutronQuery>, env: Env) -> ContractResult<Binary> {
+    let mut key = vec![];
+    let mut total_delegations = vec![];
+
+    loop {
+        let res: StdResult<QueryDelegatorDelegationsResponse> =
+            deps.querier.query(&QueryRequest::Stargate {
+                path: "/cosmos.staking.v1beta1.Query/DelegatorDelegations".to_string(),
+                data:
+                    cosmos_sdk_proto::cosmos::staking::v1beta1::QueryDelegatorDelegationsRequest {
+                        delegator_addr: env.contract.address.to_string(),
+                        pagination: Some(PageRequest {
+                            key: key.clone(),
+                            limit: 500,
+                            ..Default::default()
+                        }),
+                    }
+                    .encode_to_vec()
+                    .into(),
+            });
+
+        if res.is_err() {
+            return Ok(to_json_binary(&DelegationsResponse {
+                delegations: Delegations {
+                    delegations: vec![],
+                },
+                remote_height: 0,
+                local_height: 0,
+                timestamp: Timestamp::default(),
+            })?);
+        } else {
+            let delegations_response = res.unwrap(); // unwrap is safe bc we know that it's not an error
+
+            let delegations: Vec<DropDelegation> = delegations_response
+                .delegation_responses
+                .into_iter()
+                .map(Into::into)
+                .collect();
+
+            total_delegations.extend(delegations);
+
+            if delegations_response.pagination.next_key.is_none() {
+                break;
+            } else {
+                key = delegations_response.pagination.next_key.unwrap();
+            }
+        }
+    }
+
+    let delegations = Delegations {
+        delegations: total_delegations,
+    };
+
+    Ok(to_json_binary(&DelegationsResponse {
+        delegations,
+        remote_height: env.block.height,
+        local_height: env.block.height,
+        timestamp: env.block.time,
+    })?)
+}
+
+fn query_balances(deps: Deps<NeutronQuery>, env: Env) -> ContractResult<Binary> {
+    let balances = deps
+        .querier
+        .query_all_balances(env.contract.address.to_string())?;
```

Comment:
I am almost sure that if someone sends a lot of useless coins in different denominations to our contract, this query will consume so much gas it will never fit in a single block. I would rather query balances only in those denominations we care about.